File: framework/dotnet/BetterPortal.Tool/RegistryClient.cs
using BetterPortal;
using System.Text;
using Node = System.Collections.Generic.Dictionary<string, object?>;

namespace BetterPortal.Tool;

/// <summary>Bounded authoring requests to the existing BP contract registry.</summary>
internal sealed class RegistryClient
{
    private readonly string url;
    private static readonly UTF8Encoding Utf8 = new(false, true);
    public RegistryClient(string? url = null)
    {
        var value = url ?? Environment.GetEnvironmentVariable("BP_REGISTRY_URL") ?? "https://io.betterportal.org";
        TrustedKeys.SecureEndpoint(value); this.url = value.TrimEnd('/');
    }
    private async Task<(byte[] Data, string? Reference)> Request(string path, CancellationToken cancellation, byte[]? data = null, string? token = null)
    {
        using var deadline = CancellationTokenSource.CreateLinkedTokenSource(cancellation); deadline.CancelAfter(TimeSpan.FromSeconds(30));
        using var client = new HttpClient(new SocketsHttpHandler { AllowAutoRedirect = false, UseCookies = false, UseProxy = false }) { Timeout = Timeout.InfiniteTimeSpan };
        using var request = new HttpRequestMessage(data is null ? HttpMethod.Get : HttpMethod.Post, url + path);
        request.Headers.Accept.ParseAdd("application/json"); request.Headers.AcceptEncoding.ParseAdd("identity");
        if (token is not null)
        {
            if (token.Length is < 1 or > 32768 || token.Any(c => c is < (char)33 or > (char)126)) throw new ArgumentException("Invalid registry token");
            request.Headers.Authorization = new("Bearer", token);
        }
        if (data is not null)
        {
            if (data.Length > 16 * 1024 * 1024) throw new ArgumentException("Registry request exceeds its size limit");
            request.Content = new ByteArrayContent(data); request.Content.Headers.ContentType = new("application/json");
        }
        using var response = await client.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, deadline.Token);
        if ((int)response.StatusCode != 200 && !(data is not null && (int)response.StatusCode == 201))
            throw new ArgumentException($"Registry request failed ({(int)response.StatusCode})");
        if (!string.Equals(response.Content.Headers.ContentType?.MediaType, "application/json", StringComparison.OrdinalIgnoreCase)
            || response.Content.Headers.ContentEncoding.Any(value => !value.Equals("identity", StringComparison.OrdinalIgnoreCase)))
            throw new ArgumentException("Invalid registry response representation");
        await using var stream = await response.Content.ReadAsStreamAsync(deadline.Token);
        using var output = new MemoryStream(); var buffer = new byte[8192]; int count;
        while ((count = await stream.ReadAsync(buffer, deadline.Token)) > 0)
        {
            if (output.Length + count > 16 * 1024 * 1024) throw new ArgumentException("Registry response exceeds its size limit");
            output.Write(buffer, 0, count);
        }
        return (output.ToArray(), response.Headers.TryGetValues("BP-Registry-Ref", out var references) ? string.Join(",", references) : null);
    }
    public async Task<(string Reference, byte[] Data, Node Contract)> Lookup(string kind, string identity, string? version, CancellationToken cancellation)
    {
        if (kind == "shortName")
        {
            var (catalog, _) = await Request("/v1/packages?name=" + Uri.EscapeDataString(identity), cancellation);
            var matches = ((IEnumerable<object?>)Contracts.Parse("RegistryPackageListSchema", Json.Read(Utf8.GetString(catalog)))!).Cast<Node>().ToArray();
            if (matches.Length != 1 || ((string)matches[0]["registryRef"]!).Split('/')[^1] != identity)
                throw new ArgumentException("Registry selector is ambiguous or missing; use the full reference");
            kind = "registryRef"; identity = (string)matches[0]["registryRef"]!;
        }
        var path = kind == "registryRef" ? "/v1/packages/" + identity : "/v1/plugin-ids/" + Uri.EscapeDataString(identity);
        var (data, header) = await Request(path + "/" + Uri.EscapeDataString(version ?? "latest") + "/schema.json", cancellation);
        var reference = kind == "registryRef" ? identity : header ?? "";
        Contracts.Parse("RegistryReferenceSchema", reference);
        var contract = (Node)Contracts.Parse("BpSchemaOutputSchema", Json.Read(Utf8.GetString(data)))!; var manifest = (Node)contract["manifest"]!;
        if (kind == "pluginId" && !Equals(manifest["pluginId"], identity) || version is not (null or "latest") && !Equals(manifest["version"], version))
            throw new ArgumentException("Registry returned a different dependency identity or version");
        if (header is not null && header != reference) throw new ArgumentException("Registry returned a different package reference");
        return (reference, data, contract);
    }
    public async Task<Node> Publish(string reference, object? value, string token, CancellationToken cancellation)
    {
        Contracts.Parse("RegistryReferenceSchema", reference);
        var contract = (Node)Contracts.Parse("BpSchemaOutputSchema", value)!; var manifest = (Node)contract["manifest"]!;
        var (data, _) = await Request("/v1/packages/" + reference, cancellation, Utf8.GetBytes(Json.Write(contract)), token);
        var result = (Node)Contracts.Parse("RegistryPublishResultSchema", Json.Read(Utf8.GetString(data)))!;
        if (!Equals(result["registryRef"], reference) || !Equals(result["pluginId"], manifest["pluginId"]) || !Equals(result["version"], manifest["version"]))
            throw new ArgumentException("Registry publish response has a different identity or version");
        return result;
    }
}

File: framework/python/betterportal/registry_client.py
"""Bounded authoring requests to the existing BP contract registry."""
from __future__ import annotations

import asyncio
import json
import os
from typing import Any
from urllib.parse import quote

import httpx

from .contracts import parse
from .jsoncodec import loads
from .keys import secure_endpoint


class RegistryClient:
    def __init__(self, url: str | None = None):
        self.url = secure_endpoint(url if url is not None else os.environ.get("BP_REGISTRY_URL", "https://io.betterportal.org")).rstrip("/")

    async def _request(self, path: str, data: bytes | None = None, token: str | None = None) -> tuple[bytes, httpx.Headers]:
        async def send() -> tuple[bytes, httpx.Headers]:
            headers = {"accept": "application/json", "accept-encoding": "identity"}
            if token is not None:
                if not 1 <= len(token) <= 32768 or any(ord(c) < 33 or ord(c) > 126 for c in token):
                    raise ValueError("Invalid registry token")
                headers["authorization"] = "Bearer " + token
            if data is not None:
                if len(data) > 16 * 1024 * 1024: raise ValueError("Registry request exceeds its size limit")
                headers["content-type"] = "application/json"
            # One request per client: no cookie persistence or environment proxy credentials.
            async with httpx.AsyncClient(follow_redirects=False, trust_env=False, timeout=30) as client:
                async with client.stream("GET" if data is None else "POST", self.url + path, headers=headers, content=data) as response:
                    if response.status_code not in ((200,) if data is None else (200, 201)):
                        raise ValueError(f"Registry request failed ({response.status_code})")
                    if response.headers.get("content-type", "").split(";", 1)[0].strip().lower() != "application/json" or response.headers.get("content-encoding", "identity").lower() != "identity":
                        raise ValueError("Invalid registry response representation")
                    buffer = bytearray()
                    async for chunk in response.aiter_raw(chunk_size=8192):
                        if len(buffer) + len(chunk) > 16 * 1024 * 1024: raise ValueError("Registry response exceeds its size limit")
                        buffer.extend(chunk)
                    return bytes(buffer), response.headers
        return await asyncio.wait_for(send(), 30)

    async def lookup(self, kind: str, identity: str, version: str | None) -> tuple[str, bytes, Any]:
        if kind == "shortName":
            data, _ = await self._request("/v1/packages?name=" + quote(identity, safe=""))
            matches = parse("RegistryPackageListSchema", loads(data.decode("utf-8")))
            if len(matches) != 1 or matches[0]["registryRef"].split("/")[-1] != identity:
                raise ValueError("Registry selector is ambiguous or missing; use the full reference")
            kind, identity = "registryRef", matches[0]["registryRef"]
        path = "/v1/packages/" + identity if kind == "registryRef" else "/v1/plugin-ids/" + quote(identity, safe="")
        data, headers = await self._request(path + "/" + quote(version or "latest", safe="") + "/schema.json")
        reference = identity if kind == "registryRef" else headers.get("bp-registry-ref", "")
        parse("RegistryReferenceSchema", reference)
        contract = parse("BpSchemaOutputSchema", loads(data.decode("utf-8")))
        if kind == "pluginId" and contract["manifest"]["pluginId"] != identity or version not in (None, "latest") and contract["manifest"]["version"] != version:
            raise ValueError("Registry returned a different dependency identity or version")
        if headers.get("bp-registry-ref", reference) != reference:
            raise ValueError("Registry returned a different package reference")
        return reference, data, contract

    async def publish(self, reference: str, contract: Any, token: str) -> Any:
        parse("RegistryReferenceSchema", reference)
        contract = parse("BpSchemaOutputSchema", contract)
        data = json.dumps(contract, ensure_ascii=True, separators=(",", ":"), allow_nan=False).encode()
        response, _ = await self._request("/v1/packages/" + reference, data, token)
        result = parse("RegistryPublishResultSchema", loads(response.decode("utf-8")))
        if result["registryRef"] != reference or any(result[name] != contract["manifest"][name] for name in ("pluginId", "version")):
            raise ValueError("Registry publish response has a different identity or version")
        return result
